### DISCUSSION TITLE
Support go-to-implementation

### DISCUSSION THREAD

**Discussion by MalcolmMielle** [microsoft/pylance-release#2642] [Enhancement]
Hi all,

The "find implementation" feature doesn't seem to work well. I have a method f in a class A that inherits from B and reimplements f, and I'd like to list all "f" implementation, but nothing gets listed.
Is the feature included in Pylance, and if yes, am I missing something?

Thanks,

**Comment by jakebailey** [comment]
We don't implement the go-to or find implementation calls, so this would be an enhancement.

**Comment by judej** [comment]
Moving this issue to discussion as an enhancement request for comments and upvotes.

**Reply by yurymann** [reply]
What's the right place to upvote? Should I do it on the top post in this discussion or somewhere else?

**Reply by rchiodo** [reply]
Top of the discussion

**Comment by yurymann** [comment]
For someone like me coming from IntelliJ Idea/PyCharm, missing this features feels like a real degradation of productivity. We have a massive codebase with thousands of files and hundreds of class hierarchies. 

Now that type hierarchy has been added (#4978 - thanks!), does it make sense to also support "Find All Implementations"?

**Comment by johnnywell** [comment]
I have a couple thousand classes on my project. I've been abstracting a lot of things and also creating some protocols. But it gets tricky to understand where things came from once we make indirections with protocols. It would be great to have "Go to Implementation"!

**Comment by antoniogamizdelgado** [comment]
Same thing here!

**Reply by phenix-chen** [reply]
It is an extremely useful feature. Moreover, it is a reason that prevents me from migrating from PyCharm to Visual Studio Code.

**Comment by mrx23dot** [comment]
This is a very basic feature in IDEs, otherwise it's just a notepad.exe with dark theme.

**Reply by phenix-chen** [reply]
> I find a solution. Use "Show Type Hierarchy" to find implementation methods and "Show Calls Hierarchy" to find methods in parent classes.

try this.

**Comment by phenix-chen** [comment]
I find a solution. Use "Show Type Hierarchy" to find implementation methods and "Show Calls Hierarchy" to find methods in parent classes.

**Comment by AzMoo** [comment]
Go to implementation would be amazing. I came here because PyCharm hasn't implemented implicit implementation support for Protocols and I was coming to see if VSCode had done so, and unfortunately found I couldn't even do it for explicit implementations in VSCode.

**Comment by debonte** [comment]
Support for "Go to Implementations" shipped in 2024.12.100 -- https://github.com/microsoft/pylance-release/issues/4329

**Reply by Andrej730** [reply]
Can someone please show an example on how it works?

E.g. when I use example below and then click on `Shape.area` -> "Find All Implementations" it says "No results. Try running a previous search again".

```python
from abc import ABC, abstractmethod

class Shape(ABC):
    @abstractmethod
    def area(self) -> float:
        pass

class Circle(Shape):
    def __init__(self, radius: float):
        self.radius = radius

    def area(self) -> float:
        return 3.14 * self.radius * self.radius

class Rectangle(Shape):
    def __init__(self, width: float, height: float):
        self.width = width
        self.height = height

    def area(self) -> float:
        return self.width * self.height
```

**Reply by rchiodo** [reply]
Are you using the latest? It works for me:

![image](https://github.com/user-attachments/assets/7cd407d4-f64c-47aa-bd64-408c2628a385)

**Reply by Andrej730** [reply]
@rchiodo thank you for reply, yes. See the demo.

https://github.com/user-attachments/assets/752d99d4-f4c4-49a7-a3c3-7ef25f7262c6

```
Version: 1.99.0-insider (user setup)
Commit: 299b150bd5c0227099c9b03dc56c5073ec0ff5e2
Date: 2025-03-24T05:03:53.092Z
Electron: 34.3.2
ElectronBuildId: 11161073
Chromium: 132.0.6834.210
Node.js: 20.18.3
V8: 13.2.152.41-electron.0
OS: Windows_NT x64 10.0.22635
```

But I've got an idea what's wrong - it seems it breaks when file is outside the current workspace. Using go to implementations / find all works when file is inside workspace folder.

See another demo:
https://github.com/user-attachments/assets/0cc6ef3f-0314-4b71-b8d2-e19c1bb63ddc

**Reply by rchiodo** [reply]
Ah that makes sense. We only search the workspace for implementations. It should be smarter enough to search the current file though. I'll enter a bug.

**Comment by mrx23dot** [comment]
Thank you!!
